File: npm/src/__tests__/SOSLParserTest.ts
import {
    SoslLiteralAltContext,
    SoslLiteralContext
} from "../ApexParser";
import { createParser } from "./SyntaxErrorCounter";

test('testBasicQuery', () => {
    const [parser, errorCounter] = createParser("test.sosl", "[Find 'something' RETURNING Account]")

    const context = parser.soslLiteral()

    expect(context).toBeInstanceOf(SoslLiteralContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('testEmbeddedQuote', () => {
    const [parser, errorCounter] = createParser("test.sosl", "[Find 'some\\'thing' RETURNING Account]")

    const context = parser.soslLiteral()

    expect(context).toBeInstanceOf(SoslLiteralContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('testBracesFail', () => {
    const [parser, errorCounter] = createParser("test.sosl", "[Find {something} RETURNING Account]")

    const context = parser.soslLiteral()

    expect(context).toBeInstanceOf(SoslLiteralContext)
    expect(errorCounter.getNumErrors()).toEqual(1)
})

test('testBracesOnAltFormat', () => {
    const [parser, errorCounter] = createParser("test.sosl", "[Find {something} RETURNING Account]")

    const context = parser.soslLiteralAlt()

    expect(context).toBeInstanceOf(SoslLiteralAltContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('testQuotesFailOnAltFormat', () => {
    const [parser, errorCounter] = createParser("test.sosl", "[Find 'something' RETURNING Account]")

    const context = parser.soslLiteralAlt()

    expect(context).toBeInstanceOf(SoslLiteralAltContext)
    expect(errorCounter.getNumErrors()).toEqual(1)
})

File: jvm/src/test/java/com/nawforce/apexparser/SOSLParserTest.java
package com.nawforce.apexparser;

import org.junit.jupiter.api.Test;

import java.util.Map;

import static com.nawforce.apexparser.SyntaxErrorCounter.createParser;
import static org.junit.jupiter.api.Assertions.*;

public class SOSLParserTest {
    @Test
    void testBasicQuery() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("[Find 'something' RETURNING Account]");
        ApexParser.SoslLiteralContext context = parserAndCounter.getKey().soslLiteral();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testEmbeddedQuote() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("[Find 'some\\'thing' RETURNING Account]");
        ApexParser.SoslLiteralContext context = parserAndCounter.getKey().soslLiteral();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testBracesFail() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("[Find {something} RETURNING Account]");
        ApexParser.SoslLiteralContext context = parserAndCounter.getKey().soslLiteral();
        assertNotNull(context);
        assertEquals(1, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testBracesOnAltFormat() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("[Find {something} RETURNING Account]");
        ApexParser.SoslLiteralAltContext context = parserAndCounter.getKey().soslLiteralAlt();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testQuotesFailOnAltFormat() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("[Find 'something' RETURNING Account]");
        ApexParser.SoslLiteralAltContext context = parserAndCounter.getKey().soslLiteralAlt();
        assertNotNull(context);
        assertEquals(1, parserAndCounter.getValue().getNumErrors());
    }
}
